### PR TITLE
add resconfig comparisson and test

### LINKED ISSUE #659: Res config - constructor and test
- Update res config with dictionary
- test function that compares all the subconfigs initiated via dictionary

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves #659 

**Approach**
- inject config_dict into subsconfigs
- comparison function for res config
- test function for loading res config in two ways (config content, config dict) and compare

### CODE REVIEW

File: python/res/enkf/ensemble_config.py [hunk 1]
```diff
@@ -64,104 +64,104 @@ def __init__(self, config_content=None, grid=None, refcase=None, config_dict=Non
...
         if config_dict is not None:
-            c_ptr = self._alloc_full(config_dict[ConfigKeys.GEN_KW_TAG_FORMAT])
+            c_ptr = self._alloc_full(config_dict.get(ConfigKeys.GEN_KW_TAG_FORMAT))
             if c_ptr is None:
                 raise ValueError("Failed to construct EnsembleConfig instance from dict")
 
             super(EnsembleConfig, self).__init__(c_ptr)
 
             gen_param_list = config_dict.get(ConfigKeys.GEN_PARAM, [])
             for a in gen_param_list:
```
Comment (ManInFez) [inline python/res/enkf/ensemble_config.py]: single letter variable names should be avoided. Maybe can we change this when we already is changing the access method?

File: python/res/enkf/ert_templates.py [hunk 2]
```diff
@@ -90,7 +90,8 @@ def __ne__(self, other):
         return not self == other
 
     def __repr__(self):
-        return "ErtTemplates({})".format(", ".join(x + "=" + str(self.get_template(x)) for x in self.getTemplateNames()))
+        #return "ErtTemplates({})".format(", ".join(x + "=" + str(self.get_template(x)) for x in self.getTemplateNames()))
```
Comment (ManInFez) [inline python/res/enkf/ert_templates.py]: RMCL!

File: python/res/enkf/ert_workflow_list.py [hunk 3]
```diff
@@ -41,12 +41,17 @@ def __init__(self, subst_list=None, config_content=None, config_dict=None):
             parser = WorkflowJob.configParser( )
             for job in config_dict.get(ConfigKeys.LOAD_WORKFLOW_JOB, []):
                 try:
-                    new_job = WorkflowJob.fromFile(config_file=job['PATH'], name=job['NAME'], parser=parser)
+                    new_job = WorkflowJob.fromFile(config_file=job[ConfigKeys.PATH], name=job[ConfigKeys.NAME], parser=parser)
+                    print('DOING: {} {}'.format(job[ConfigKeys.PATH],job[ConfigKeys.NAME]))
```
Comment (ManInFez) [inline python/res/enkf/ert_workflow_list.py]: Remove print statement

File: python/res/enkf/ert_workflow_list.py [hunk 4]
```diff
@@ -41,12 +41,17 @@ def __init__(self, subst_list=None, config_content=None, config_dict=None):
...
                 except:
-                    print("WARNING: Unable to create job from {}".format(job['PATH']))
+                    print("WARNING: Unable to create job from {}".format(job[ConfigKeys.PATH]))
                     continue
-                workflow_joblist.addJob(new_job)
-                new_job.convertToCReference(None)
+                if not new_job is None:
+                    workflow_joblist.addJob(new_job)
+                    new_job.convertToCReference(None)
+                    print('ADD: {} {}'.format(job[ConfigKeys.PATH], job[ConfigKeys.NAME]))
```
Comment (ManInFez) [inline python/res/enkf/ert_workflow_list.py]: Remove print statement

File: python/res/enkf/ert_workflow_list.py [hunk 5]
```diff
@@ -98,7 +103,8 @@ def getContext(self):
         return self._get_context()
 
     def __str__(self):
-        return 'ErtWorkflowList with jobs: %s' + str(self.getJobNames())
+        return 'ErtWorkflowList'
+        # return 'ErtWorkflowList with jobs: %s' + str(self.getJobNames())
```
Comment (ManInFez) [inline python/res/enkf/ert_workflow_list.py]: Why not return the job names?

File: python/res/enkf/hook_manager.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ def __init__(self, workflow_list, config_content=None, config_dict=None):
 
             # RUNPATH_FILE #
             runpath_file_name = config_dict.get(ConfigKeys.RUNPATH_FILE, ConfigKeys.RUNPATH_LIST_FILE)
-            runpath_file_path = os.path.join(config_dir, runpath_file_name)
+            runpath_file_path = os.path.normpath(os.path.join(config_dir, runpath_file_name))
```
Comment (ManInFez) [inline python/res/enkf/hook_manager.py:40]: why not realpath?

File: python/res/enkf/model_config.py [hunk 7]
```diff
@@ -106,17 +106,22 @@ def __init__(self, data_root, joblist, last_history_restart, refcase,
...
-                job = forward_model.add_job(job_description['NAME'])
-                args = StringList(job_description['ARGUMENTS'])                
-                job.set_args(args)
-                args.convertToCReference(None)
+            # SIMULATION_JOB_KEY
+            for job_description in config_dict.get(ConfigKeys.FORWARD_MODEL, []):
+                job = forward_model.add_job(job_description[ConfigKeys.NAME])
+                job.set_private_args_as_string(job_description.get(ConfigKeys.ARGLIST))
                 job.convertToCReference(None)
+            # for job_description in config_dict.get(ConfigKeys.FORWARD_MODEL, []):
```
Comment (ManInFez) [inline python/res/enkf/model_config.py]: Remove commented code

File: python/res/enkf/queue_config.py [hunk 8]
```diff
@@ -57,11 +57,11 @@ def __init__(self, user_config_file=None, config_content=None, config_dict=None)
 
         if config_dict is not None:
                 c_ptr = self._alloc_full(
-                                            config_dict[ConfigKeys.JOB_SCRIPT],
-                                            config_dict[ConfigKeys.USER_MODE],
-                                            config_dict[ConfigKeys.MAX_SUBMIT],
-                                            config_dict[ConfigKeys.NUM_CPU],
-                                            config_dict[ConfigKeys.QUEUE_SYSTEM]
+                                            config_dict.get(ConfigKeys.JOB_SCRIPT),
```
Comment (ManInFez) [inline python/res/enkf/queue_config.py]: None of these values can, or should be None for ERT to function

File: python/res/enkf/res_config.py [hunk 9]
```diff
@@ -66,53 +66,76 @@ class ResConfig(BaseCClass):
     _init_parser       = ResPrototype("void res_config_init_config_parser(config_parser)", bind=False)
 
 
-    def __init__(self, user_config_file=None, config=None, throw_on_error=True):
+    def __init__(self, user_config_file=None, config=None, throw_on_error=True, config_dict=None):
 
+        #for backward compatibility, needs to be supplied in everest as config_dict instead of config
+        if config is not None:
```
Comment (ManInFez) [inline python/res/enkf/res_config.py]: Why is this needed. Can't we just overtake the usage of config?

File: python/res/enkf/subst_config.py [hunk 10]
```diff
@@ -51,23 +51,23 @@ def __init__(self, config_content=None, config_dict=None):
...
+            # DATA_KW
             data_kw = config_dict.get(ConfigKeys.DATA_KW_KEY)
             if isinstance(data_kw, dict):
                 for key, value in data_kw.items():
                     subst_list.addItem(key, value)
 
             # RUNPATH_FILE #
             runpath_file_name = config_dict.get(ConfigKeys.RUNPATH_FILE, ConfigKeys.RUNPATH_LIST_FILE)
-            runpath_file_path = os.path.join(config_directory, runpath_file_name)
+            runpath_file_path = os.path.normpath(os.path.join(config_directory, runpath_file_name))
```
Comment (ManInFez) [inline python/res/enkf/subst_config.py:70]: Why normapath?

File: python/res/enkf/ert_templates.py (hunk 2 above)
Comment (xjules) [reply python/res/enkf/ert_templates.py]: OK

File: python/res/enkf/ensemble_config.py (hunk 1 above)
Comment (xjules) [reply python/res/enkf/ensemble_config.py]: Which single letter variable you mean actually - c_ptr?

File: python/res/enkf/ert_workflow_list.py (hunk 5 above)
Comment (xjules) [reply python/res/enkf/ert_workflow_list.py]: That was due to debugger - now putting them back

File: python/res/enkf/hook_manager.py (hunk 6 above)
Comment (xjules) [reply python/res/enkf/hook_manager.py:40]: it compares to abspath. Also when joining two paths might produce `.../...`, which fails in tests.

File: python/res/enkf/queue_config.py (hunk 8 above)
Comment (xjules) [reply python/res/enkf/queue_config.py]: Ok,  reverting it back

File: python/res/enkf/model_config.py (hunk 7 above)
Comment (xjules) [reply python/res/enkf/model_config.py]: ok

File: python/res/enkf/ert_workflow_list.py (hunk 3 above)
Comment (xjules) [reply python/res/enkf/ert_workflow_list.py]: ok

File: python/res/enkf/res_config.py (hunk 9 above)
Comment (xjules) [reply python/res/enkf/res_config.py]: It's until we make everest ready, since it uses `build_config_content`hence has a bit different structure.

File: python/res/enkf/subst_config.py (hunk 10 above)
Comment (xjules) [reply python/res/enkf/subst_config.py:70]: The same as before: it compares to abspath. Also when joining two paths might produce `.../...`, which fails in tests.

Comment (ManInFez) [reply python/res/enkf/subst_config.py:70]: :+1:

File: python/res/enkf/res_config.py (hunk 9 above)
Comment (ManInFez) [reply python/res/enkf/res_config.py]: Then maybe restructure the init function and split into three sub functions. config_file, config, and config_dict. It may be result in a bit of redundant code, but will make it really a lot more readable

Comment (xjules) [reply python/res/enkf/res_config.py]: ok, have a look whether it is clear enough

**Comment by ManInFez** [comment]
test this please

**Review by ManInFez — APPROVED** [review]
God job! :+1:
